src/environments/array.js

```diff
@@ -934,6 +947,36 @@ defineEnvironment({
     mathmlBuilder,
 });
 
+defineEnvironment({
+    type: "array",
+    names: ["equation", "equation*"],
+    props: {
+        numArgs: 0,
+    },
+    handler(context) {
+        const settings = context.parser.settings;
+        if (!settings.displayMode) {
+            throw new ParseError(`{${context.envName}} cannot be used inline.`);
+        } else if (settings.strict && !settings.topEnv) {
+            settings.reportNonstrict("textEnv",
+                `{${context.envName}} called from math mode.`);
+        }
+        const res = {
+            addEqnNum: context.envName === "equation",
+            maxNumRows: 1,
+            maxNumCols: 1,
+            leqno: context.parser.settings.leqno,
+        };
+        const equation = parseArray(context.parser, res, "display");
+        if (equation.body.length > 1) {
+            throw new ParseError("Misplaced \\cr.");
```

Comment:
This also can be move to `parseArray`. Sorry for spread reviews, I didn't notice this at first.

Reply:
It can move into `parseArray` but I don't know if it should. If it is is inside `parseArray`, then every environment does that check. As currently coded, only `{equation}` and `{equation*}` do that check. One method of getting good performance is to put less code into the inner loop.

Reply:
I don't think a simple truthiness check will affect much performance. IMHO, I think it's better to fail early with the token information.

Reply:
Done.